darwinpush/ftp.py

```diff
@@ -0,0 +1,189 @@
+import ftplib
+import datetime
+import re
+
+import pyxb.utils.domutils as domutils
+
+import logging
+log = logging.getLogger("ftp")
+
+ftp_server = "datafeeds.nationalrail.co.uk"
+
+# Downtime required to need logs from Darwin.
+log_threshold = datetime.timedelta(minutes=5)
+
+# Downtime required to need a snapshot.
+snapshot_threshold = datetime.timedelta(days=1)
+
+# A downtime <=0 downlods the snapshot and all the logs.
+
+
+_snapshot_pattern = re.compile("^([0-9]{4})([0-9]{2})([0-9]{2}).*_v8\.xml\.gz$")
+_log_pattern = re.compile(
+    "^pPortData.log.([0-9]{4})-([0-9]{2})-([0-9]{2})-([0-9]{2})-([0-9]{2})$"
+)
+
+class LoginFail(Exception): pass
+
+
+def fetchAll(client, downtime, server=ftp_server, user=None, passwd=None):
+    """Download all the FTP files required, parse them and pass them to the
+    client.
+
+    It connects to the FTP server, downloads the required files and passes the
+    messages to the client, and disconnects from the FTP server.
+
+    This method blocks until it's done processing.
+
+    Args:
+        client: Must have a on_message(headers, message, source) method, which
+              will be called for every valid message parsed.
+
+        downtime: An int representing the number of seconds of downtime. It
+                can also be a datetime.timedelta representing the downtime.
+
+                If the number of seconds is:
+                    <=0, then the snapshot for the day will be downloaded
+                         and applied, and also all the logs.
+
+                         NOT YET IMPLEMENTED.
+
+                    >0,  all the required logs are downloaded. This means no
+                         logs if less than 5 min (300 s) downtime, as Darwin
+                         holds 5 minutes of messages in the queue before it
+                         pushes the log to the FTP server and removes the
+                         messages from the waiting queue.
+
+                When the files from FTP are parsed, only the messages that
+                are timestamped by darwin as being sent starting from
+                `current_time - downtime` will be sent to the listener.
+
+        server: FTP server to use. Default: "datafeeds.nationailrail.co.uk".
+        user: FTP username, default None.
+        passwd: FTP password, default None.
+    """
+
+    if type(downtime) is not datetime.timedelta:
+        downtime = datetime.timedelta(seconds=downtime)
+
+    #snapshot = downtime.total_seconds() <= 0 or downtime > snapshot_threshold
```

Comment:
Can go away. It's what I started doing for snapshots. 

Same for next similar ones.